### ISSUE TITLE
Rate limit requests to Compact Index API

### ISSUE THREAD

**Issue by deivid-rodriguez** [rubygems/rubygems#3594]
The compact index API has some rate limits that bundler in very rare occassions reaches when requesting dependency information for resolution.

In that case, bundler falls back to requesting the same information, but sequentially, not with parallel threads, so that it can get it without getting TooManyRequest responses.

But in some even more rare cases, the API limits are still hit even after that fallback. If that happens, bundler falls back again to using the old dependency API, which actually has different semantics from the new API, since it doesn't provide the `required_ruby_version` field.

This leads, in very very rare occasions, to bundler incorrectly resolving and crashing at installation time.

I believe we can solve this by rate limiting the requests bundler does to the API, accomodating the number of requests bundler does to the limits set on the server side.

By doing this, we could get rid of all the fallbacks and eventually start phasing out the old dependency API.

I found this little cute gem which is basically two simple ruby files which we could use: https://github.com/Shopify/limite.

Related to https://github.com/rubygems/rubygems/issues/3463.

**Comment by simi** [comment]
:information_source: https://github.com/Shopify/limite 404s for me

**Comment by deivid-rodriguez** [comment]
Typo! Thanks!

**Comment by deivid-rodriguez** [comment]
After discussion with @indirect, we agree that what would make most sense here is to use the `Retry-After` response header and sleep for the suggested number of seconds.

**Comment by indirect** [comment]
yep! and if we get a 429 with no retry-after, waiting 1-2s and trying again seems better than falling back to the v1 api 👍🏻

**Comment by pirj** [comment]
Bump.
We have had an enormous number of weird flaky failures in RSpec CI jobs, and had to apply those workarounds:
https://github.com/rspec/rspec-rails/pull/2360/files#diff-8c3955be04b733dbd31e54e50844fb84R62
https://github.com/rspec/rspec-rails/pull/2366/files#diff-a8a7f2d3f06e0c6f9a7717dd8c7cb715R38

Indeed, there seems to be a correlation between the Bundler API being overload and the amount of Ruby version dependency related failures in build jobs. Unfortunately, I didn't keep track of those failures, but there were a couple of failures due to rubygems.org timeouts I've noticed in the logs during the past few days.

Highly appreciate if you could move this forward. It would be insanely pleasant to remove those hacks from RSpec's codebase and get rid of build flakiness once and for all.
